lib/rabbitmqctl.ex

```diff
@@ -67,13 +67,52 @@ defmodule RabbitMQCtl do
       _ ->
         options = parsed_options |> merge_all_defaults |> normalize_options
         {arguments, options} = command.merge_defaults(arguments, options)
+
         with_distribution(options, fn() ->
-          validate_and_run_command(options, command, arguments)
-          |> handle_command_output(command, options, unparsed_command, output_fun)
+          # The code below implements a tiny decision tree that has
+          # to do with CLI argument and environment state validation.
+
+          # validate CLI arguments
+          case command.validate(arguments, options) do
+            :ok ->
+              # then optionally validate execution environment
+              case maybe_validate_execution_environment(command, arguments, options) do
+                :ok   -> proceed_to_execution(command, arguments, options)
+                other -> other
+              end
+            other -> other
+            # handle_command_output will handle all the errors
+            # among other things
+          end |> handle_command_output(command, options, unparsed_command, output_fun)
```

Comment:
If we have `validation_failure` from both validation steps, how do we differentiate which message should we format? I'm talking about "Given: ... Usage: ..." message, which does not make much sense for `validate_execution_environment` errors.